File: lg/src/questions/q242_valid_anagram.go
package questions

func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	sMap := make(map[rune]uint)

	for _, c := range s {
		if v, ok := sMap[c]; ok {
			sMap[c] = v + 1
		} else {
			sMap[c] = 1
		}
	}

	for _, c := range t {
		if v, ok := sMap[c]; ok {
			if v == 1 {
				delete(sMap, c)
			} else {
				sMap[c] = v - 1
			}
		}
	}

	return len(sMap) == 0
}

File: lt/src/q-242-valid-anagram.ts
export function isAnagram(s: string, t: string): boolean {
  if (s.length != t.length) {
    return false;
  }

  const s_map = new Map<string, number>();

  s.split("").forEach((c) => {
    if (s_map.has(c)) {
      s_map.set(c, s_map.get(c)! + 1);
    } else {
      s_map.set(c, 1);
    }
  });

  t.split("").forEach((c) => {
    if (s_map.has(c)) {
      const v = s_map.get(c)!;
      if (v == 1) {
        s_map.delete(c);
      } else {
        s_map.set(c, v - 1);
      }
    }
  });

  return s_map.size == 0;
}
